src/conn_sock.c

```diff
@@ -111,51 +145,150 @@ char *setup_attach_socket(void)
 	if (listen(attach_socket_fd, 10) == -1)
 		pexitf("Failed to listen on attach socket: %s", attach_sock_path);
 
-	g_unix_fd_add(attach_socket_fd, G_IO_IN, attach_cb, NULL);
+	g_unix_fd_add(attach_socket_fd, G_IO_IN, attach_cb, &rmt_attach_sock);
 
 	return attach_symlink_dir_path;
 }
 
-static gboolean attach_cb(int fd, G_GNUC_UNUSED GIOCondition condition, G_GNUC_UNUSED gpointer user_data)
+void setup_notify_socket(char *socket_path)
+{
+	int notify_socket_fd = -1;
+	struct sockaddr_un notify_addr = {0};
+	notify_addr.sun_family = AF_UNIX;
+	char cwd[1024];
+
+	/* Connect to Host socket */
+	if (lcl_notify_host_fd < 0) {
+		lcl_notify_host_fd = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
+		if (lcl_notify_host_fd == -1) {
+			pexit("Failed to create notify socket");
+		}
+		lcl_notify_host_addr.sun_family = AF_UNIX;
+		strncpy(lcl_notify_host_addr.sun_path, socket_path, sizeof(lcl_notify_host_addr.sun_path) - 1);
+	}
+
+	/*
+	 * Create a symlink so we don't exceed unix domain socket
+	 * path length limit.
+	 */
+	_cleanup_free_ char *notify_symlink_dir_path = g_build_filename(opt_socket_path, opt_cuuid, NULL);
+	if (unlink(notify_symlink_dir_path) == -1 && errno != ENOENT)
+		pexit("Failed to remove existing symlink for notify socket directory");
+
+	/*
+	 * This is to address a corner case where the symlink path length can end up being
+	 * the same as the socket.  When it happens, the symlink prevents the socket from being
+	 * be created.  This could still be a problem with other containers, but it is safe
+	 * to assume the CUUIDs don't change length in the same directory.  As a workaround,
+	 *  in such case, make the symlink one char shorter.
+	 */
+	if (strlen(notify_symlink_dir_path) == (sizeof(notify_addr.sun_path) - 1))
+		notify_symlink_dir_path[sizeof(notify_addr.sun_path) - 2] = '\0';
+
+	if (symlink(opt_bundle_path, notify_symlink_dir_path) == -1)
+		pexit("Failed to create symlink for notify socket");
```

Comment:
we could create a helper function with this, as I'm pretty sure it's duplicated with setup_console_socket

Reply:
Slightly different, since we chdir - Let me look closer.

Reply:
I'd be in favor of pulling out the common bits to reduce duplication

Reply:
Dun